### DISCUSSION TITLE
2023-09-22 - Testing 5 - Do you get a human to test? - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by piers-sinclair** [SSWConsulting/SSW.Rules.Content#6789] [Chewing the Fat]
Hey Testers,

Let's talk about this rule | [Do you conduct a "Test Please"?](https://www.ssw.com.au/rules/conduct-a-test-please/)

**Comment by MattParkerDev** [comment]
I think we should have an SSW Rule for human testing if there is expectation or an agreed upon standard for PBI test pleases other than PRs :)

**Reply by leehawkinsssw** [reply]
These rules might help:

https://www.ssw.com.au/rules/what-is-exploratory-testing/
https://www.ssw.com.au/rules/enough-testing/
https://www.ssw.com.au/rules/how-to-decide-what-to-test/
https://www.ssw.com.au/rules/know-when-you-have-found-a-problem/

**Comment by JackDuan123** [comment]
Code coverage is just how much code is covered, even if all the branches are covered, a lot of the cases is not covered.

Let me ask you a simple question:

how can you test all the poosible out put of "+" operator?

The answer is you can never do it.

That's the importance of huam test, and that's why the world of computer programming is almost a magic world, because it running smoothly without fully tested :)

**Reply by ozdemir-mehmet** [reply]
Yep **human** testing is important :)

**Comment by ricksu978** [comment]
100% Test Please!
Human testers can cross-reference features against acceptance criteria, which may require a nuanced understanding.

**Comment by 2fernandez** [comment]
❤️The video - Test pass from designers.

The video - Awesome presenter 
The video - great camera work, all footage was in focus
The Audio - nice and clear

The rule - Nice recap on a super valuable rule

Now I'm going to try a test please with the wife, wish me luck.
😝

**Comment by matt-goldman** [comment]
The rule and video aptly and concisely capture and convey a simple concept: we are a team.

The value of a second pair of eyes on your work cannot be overstated. It's the best way to get unblocked, and it's also the best way to ensure you haven't missed anything. It's all too easy to get swept up in the minutiae of a particular detail and miss something else. The rule and video state simply that no matter what you're working on, there's always a good way to get someone else to take a look. It's a microcosm of why commercial software and games don't ship after only alpha testing, and it's probably the most important tool in your kit to prevent embarrassment in front of clients.

**Comment by brydeno** [comment]
In my view the rule lacks any content suggesting anything more than a review of code. We need to encourage actually looking at the delivered system.

**Comment by SamBWagner** [comment]
It's imperative that an over-the-shoulder review happens. There are countless opportunities for criticism that would be otherwise lost in a PR to happen and when the person is in a call/next to you, they are more receptive to helping and you're more receptive to receiving that help.

**Comment by tkapa** [comment]
Test pleasing is valuable. Hard to get done when a client is remote only. However, internal test pleasing should be done constantly.

**Comment by leehawkinsssw** [comment]
The simplification and modernization of this rule was overdue, so it now more accurately reflects how we work. I would encourage those performing "Test Please" via reviewing PRs to also experience that PR within the context of using the software, not just from the perspective of whether the code looks good.

(Please please don't blame testers for bugs that slip through, remember they didn't create them and they're also not solely responsible for identifying them)

**Comment by JackDevAU** [comment]
As they say, you can never test enough 😅 
Getting a "test please" is a super powerful and should be done whenever possible.

**Reply by leehawkinsssw** [reply]
If you're struggling to work out what "enough" looks like when you're doing any testing, this rule might help:
https://www.ssw.com.au/rules/enough-testing/

**Comment by apforsyth** [comment]
Having a second set of eyes on your work is super valuable!

**Comment by jakebayliss** [comment]
One of the most valuable SSW rules and so important for keeping ourselves to a high standard. 
I like the idea of pulling the PR and running any automated tests against code being merged. Then doing any human tests via steps the author provides

**Comment by ozdemir-mehmet** [comment]
Great to get an extra set of eyes on your work (when time permits)

**Comment by AlexRothwell** [comment]
I find it super valuable to get my PBIs checked by someone else as they'll come at it from a different perspective. It helps to make sure I deliver the work wanted by the project

**Comment by MarcusIrmscherSSW** [comment]
Great rule, it can be applied across all aspects.

**Comment by AshAnil** [comment]
Great video ⭐ It is always better to get Checked by xxx.

**Comment by joshbermanssw** [comment]
Always good to get a checked by when others are free.

**Comment by UlyssesMaclaren** [comment]
"Checked By" FTW :)

**Comment by vladislav-kir** [comment]
It is super important to have a second pair of eyes check your work :)

**Comment by jeoffreyfischer** [comment]
Quick and easy rule !

**Comment by Levijj22** [comment]
"Test please" is super valuable. Some people might feel like its a lack of trust, especially when you're new to the company, however once you view it as a safeguard and realise it instills a culture of "protection", it will be your best friend!

**Comment by jackreimers** [comment]
This rule and video was great. It was concise and communicated the key points really well

**Comment by Ben0189** [comment]
PR is necessary for any code changes. If pr reviewer have time, over the shoulder review is recommended

**Comment by BenDunkerleySSW** [comment]
"test please" and its many mediums are invaluable no matter what the task. It can sometimes be confronting, but it almost always leads to better work!

**Comment by AttackOnMorty** [comment]
Next step: [Pair Programming](https://martinfowler.com/articles/on-pair-programming.html) (Test please during coding)

**Comment by Jord-Gui** [comment]
Great system to ensure I minimise any issues in my work.

**Comment by wicksipedia** [comment]
Don't forget that testing during development isn't just about the code - it should start when the requirements are being written! It's much cheaper to fix a bug in a requirement than later down the track

**Comment by christoment** [comment]
Love test pleases combined with the Michael Jackson rule gives your work more credibility and less mistakes 😄

**Comment by chrisschultzssw** [comment]
Most of these checks/tests I am doing (or asking for) are for emails or documentation. This is still super important, especially if it is going to a client - a different point of view can be a valuable resource.

**Comment by Harry-Ross** [comment]
Really invaluable rule and video that succinctly described the importance of "Test Please" and the importance of human testing. 

I believe that human testing in web apps is incredibly important, as sometimes it can be challenging to automatedly test unexpected behaviour by users.

**Comment by danielmackay** [comment]
Great video @william-liebenberg!  Great overview and reminder of the different types of 'test pleases' that we should all be doing.

**Comment by tomek-i** [comment]
Human testing > automated testing or Human testing < automated testing .... 

I do not think that one is more and the other less important ... automated testing is a tool that helps you with your task at hand ... 
It's comparing handwashing your laundry or using a washing machine ...  yes, there are and will be cases and situations that need delicacy.

In short, you should use the right tool for the right job - make your life easy :-)

**Comment by Abf09** [comment]
Feedback always helps in making a better version as long as its constructive. Using checked by on the emails and PR reviews have always improved my emails and PRs.

**Comment by sethdaily** [comment]
(Checked by Chris H)
Low effort, high importance.

**Comment by bahjatalaadelssw** [comment]
It's a good way to encourage accountability, critical feedback and team engagement.

**Comment by steven0x51** [comment]
"Test please" is valuable for all kink of jobs, and it keeps us high standard

**Comment by ozair-ashfaque** [comment]
Test Please is a great rule and it improves the integrity of content

**Comment by amankumarrr** [comment]
Human testing provides an opportunity to grasp the context and significance of the task.

**Comment by jomelsontanco** [comment]
✅ Rule highlighted that a Test Please is more than just a Pull Request.
❌ Rule should give guidelines on how to promote a culture of Test Please beyond coding and email "Checked by".
❌ Rule should give guidelines on how to Test Please for processes involved **prior** a Pull Request i.e. architecture considerations, component design.

**Comment by tombui99** [comment]
It is always important to have a second eyes on your work to improve quality and minimize errors

**Comment by drwharris** [comment]
Im always amazed at the objections to human/manual testing.

**Comment by stephanfakossw** [comment]
Great rule...great video...brilliant haircut - you should bring it back Uly 😊

**Comment by CoffeeBugHub** [comment]
Like 'test please!' It is beneficial to get another pair of eyes to spot the any possible problems earlier.

**Comment by PennyWalker** [comment]
Loved the video - great recap, really well shot and loved seeing some new faces in videos. Love the Rule - I think human testing is important, it's so easy to miss little mistakes that fresh eyes also catch. I'm a big believer in the test please/checked by :)

**Comment by Dhruv-0987** [comment]
Video was amazing, short and to the point. IMO you can only write tests so far specially when it comes to functional testing or usability testing which is the public facing component of the product and is the most essential. Without human testing its impossible.

**Comment by Hona** [comment]
It seems we're starting to add lots of Decision Trees to rules.
I wonder if we should get a common way to do them?

To me the major piece of this rule is the links away to test please/checked by etc, but get hidden under a tiny A tag.

CC: @jaydenalchin @matt-goldman @william-liebenberg

**Comment by YalalCakirSSW** [comment]
Getting someone else's opinion with "Test Please" can be helpful in finding details you may be missing.

**Comment by jernejk** [comment]
In 2016, I introduced SSW to pull requests in SugarLearning and disposable environments per PR. :)
(thanks to Adam Stephensen and other Brissy guys, git as well as pull requests were widely adopted)

You might notice we don't have disposable environments, as they were slow to provision and we could only have 3 at a time. (last 3 PRs and then we start to kill old ones and create new ones)

It was great for testing and for a while we managed to catch bugs before they hit to staging and even let stakeholders test it before it even hit the `master` branch. But it had its own issues as it wasn't perfectly close to the original staging environment and was later removed.

In short, it's awesome for manual testing if done right, AutoScrum for instance has an improved version of this! :) (yes, when you create a new PR, you can test the app without needing to pull the PR!)

**Comment by jerryluossw** [comment]
Great rule, There is no substitute for manual testing, users always use the system in different ways.

**Comment by Calvin-Lieu** [comment]
As a student who has to work on projects solo for the most part, I can definitely see how helpful it would be to have another person helping to review my code from an outside perspective.

**Comment by babakamyljanovssw** [comment]
Pull Requests are very important, and I took many valuable feedbacks from other developers that improved the important things I was missing.

**Comment by sylhuang** [comment]
Manual test is time-consuming but without it many mistakes cannot be caught.

**Comment by AnkyChe** [comment]
Love the rule. Manual testing is important as it can discover problems.

**Comment by Aibono1225** [comment]
Human testing is important, as it can provide double assurance for tasks, and others can often find issues that may have been overlooked.

**Comment by KristenHu** [comment]
Manual testing may not always be necessary as it may not always uncover issues. However, it can significantly save time when problems are identified.

**Comment by leahy268** [comment]
Human testing is super important.  It's really simple for a human to pick on something that is almost impossible to test automatically.  E.G.  When opening a website I could easily see if someone had mistakenly used 2 different shades of red for a banner.  However testing that automatically would be a lot harder.

**Comment by zacharykeeping** [comment]
Good, simple rule and video that's definitely best to abide by.

**Comment by BrookJeynes** [comment]
Test please is always a great way of getting a second set of eyes to check over what you've done

**Comment by tino-liu** [comment]
Let's put ourselves in other people's shoes and think. Great rule!

**Comment by JeanThirion** [comment]
Test Please are awesome. I used to do that a lot pre-SSW, but I like how we formalize the process a bit more. I love when I get "Test please" calls or emails <3

**Comment by cvionnet** [comment]
Great video, really enjoy the pace.
Looking forward to Tylah' talk to learn more about Vercel!

**Comment by jimmidier** [comment]
Human testing has the ability to inspect from the system's perspective and improvise to meet requirements, whereas automated testing done by programs may only have the knowledge within the particular portion of context and not able to cater for changes.

**Comment by Marxoz** [comment]
Automated tests are great for any project, but it is really hard to cover all scenarios, especially when end users are applicable. Manual testing will always have its place in testing.

I always explain to clients that the cost of fixing something in production is always higher than having a dedicated tester to pick up issues.
